src/org/opendatakit/briefcase/export/ExportForms.java

```diff
@@ -37,36 +38,53 @@
   private final List<BiConsumer<String, LocalDateTime>> onSuccessfulExportCallbacks = new ArrayList<>();
   private Map<String, FormStatus> formsIndex = new HashMap<>();
 
-  public ExportForms(List<FormStatus> forms, ExportConfiguration defaultConfiguration, Map<String, ExportConfiguration> configurations, Map<String, LocalDateTime> lastExportDateTimes) {
+  public ExportForms(List<FormStatus> forms, ExportConfiguration defaultConfiguration, Map<String, ExportConfiguration> configurations, Map<String, LocalDateTime> lastExportDateTimes, Map<String, ServerConnectionInfo> transferSettings) {
     this.forms = forms;
     this.defaultConfiguration = defaultConfiguration;
     this.customConfigurations = configurations;
     this.lastExportDateTimes = lastExportDateTimes;
+    this.transferSettings = transferSettings;
     rebuildIndex();
   }
 
-  public static ExportForms load(ExportConfiguration defaultConfiguration, List<FormStatus> forms, BriefcasePreferences preferences) {
+  public static ExportForms load(ExportConfiguration defaultConfiguration, List<FormStatus> forms, BriefcasePreferences exportPreferences, BriefcasePreferences appPreferences) {
     // This should be a simple Map filtering block but we'll have to wait for Vavr.io
 
     Map<String, ExportConfiguration> configurations = new HashMap<>();
     Map<String, LocalDateTime> lastExportDateTimes = new HashMap<>();
+    Map<String, ServerConnectionInfo> transferSettings = new HashMap<>();
     forms.forEach(form -> {
       String formId = getFormId(form);
-      ExportConfiguration load = ExportConfiguration.load(preferences, buildCustomConfPrefix(formId));
+      ExportConfiguration load = ExportConfiguration.load(exportPreferences, buildCustomConfPrefix(formId));
       if (!load.isEmpty())
         configurations.put(formId, load);
-      preferences.nullSafeGet(buildExportDateTimePrefix(formId))
+      exportPreferences.nullSafeGet(buildExportDateTimePrefix(formId))
           .map(LocalDateTime::parse)
           .ifPresent(dateTime -> lastExportDateTimes.put(formId, dateTime));
+      String urlKey = String.format("%s_pull_settings_url", formId);
+      String usernameKey = String.format("%s_pull_settings_username", formId);
+      String passwordKey = String.format("%s_pull_settings_password", formId);
+      if (appPreferences.hasKey(urlKey) && appPreferences.hasKey(usernameKey) && appPreferences.hasKey(passwordKey))
+        transferSettings.put(formId, new ServerConnectionInfo(
+            appPreferences.nullSafeGet(urlKey)
+                .orElseThrow(() -> new RuntimeException("Null value saved for " + urlKey)),
+            appPreferences.nullSafeGet(usernameKey)
+                .orElseThrow(() -> new RuntimeException("Null value saved for " + usernameKey)),
+            appPreferences.nullSafeGet(passwordKey)
+                .orElseThrow(() -> new RuntimeException("Null value saved for " + passwordKey)).toCharArray()
+        ));
+
```

Comment:
I think this reads so much better:
```
transferSettings.put(formId, new ServerConnectionInfo(
    getPrefOrThrow(appPreferences, urlKey),
    getPrefOrThrow(appPreferences, usernameKey),
    getPrefOrThrow(appPreferences, passwordKey).toCharArray()
));

  private static String getPrefOrThrow(BriefcasePreferences appPreferences, String key) {
    return appPreferences.nullSafeGet(key).orElseThrow(() -> new RuntimeException("Null value saved for " + key));
  }
```
What do you think?

Reply:
I was waiting to refactor `BriefcasePreferences.get(key)` to make it throw when the key doesn't exist and rename `nullSafeGet(key)` to `getOption(key)`. With these changes it would become: 

```java
transferSettings.put(formId, new ServerConnectionInfo(
    appPreferences.get(urlKey),
    appPreferences.get(usernameKey),
    appPreferences.get(passwordKey).toCharArray()
));
```

Reply:
Even better!